include/sparrow/layout/primitive_array.hpp

```diff
@@ -155,6 +192,120 @@ namespace sparrow
         SPARROW_ASSERT_TRUE(get_arrow_proxy().data_type() == arrow_traits<T>::type_id);
     }
 
+    template <class T>
+    template <std::ranges::input_range VALUE_RANGE, std::ranges::input_range BOOL_RANGE>
+    requires std::convertible_to<std::ranges::range_value_t<VALUE_RANGE>, T> &&
+             std::convertible_to<std::ranges::range_value_t<BOOL_RANGE>, bool>
+    arrow_proxy primitive_array<T>::create_proxy(VALUE_RANGE && values, BOOL_RANGE && is_non_null)
+    {
+
+        const auto range_size = std::ranges::size(values);
+
+        // buffers
+        auto bitmap_ptr = new std::uint8_t[(range_size + 7) / 8];
+        auto data_buffer_ptr = new std::uint8_t[range_size * sizeof(T)];
+
+        #ifdef __GNUC__
+        #    pragma GCC diagnostic push
+        #    pragma GCC diagnostic ignored "-Wcast-align"
+        #endif
+        auto data_ptr = reinterpret_cast<T*>(data_buffer_ptr);
+        #ifdef __GNUC__
+        #    pragma GCC diagnostic pop
+        #endif
+
+        // iterators to the inputs
+        auto value_iter = std::ranges::begin(values);
+        auto is_non_null_iter = std::ranges::begin(is_non_null);
+
+        // fill the buffers
+        for(std::size_t i=0; i < range_size; ++i)
+        {
+            const bool is_non_null_val = *is_non_null_iter;
+            if(is_non_null_val)
+            {
+                data_ptr[i] = static_cast<T>(*value_iter);
+                // set bit to 1
+                bitmap_ptr[i / 8] |= static_cast<std::uint8_t>(1 << (i % 8));
+            }
+            else
+            {
+                // set bit to 0
+                bitmap_ptr[i / 8] &= static_cast<std::uint8_t>(~(1 << (i % 8)));
+            }
+            ++value_iter;
+            ++is_non_null_iter;
+        }
```

Comment:
Filling the bitmap should be done in a dedicated method (notice it is alrady implemented [here](https://github.com/man-group/sparrow/blob/main/include/sparrow/arrow_interface/arrow_array_schema_factory.hpp#L57-L71)) to avoid code duplictation.

Reply:
@JohanMabille the function you show does something different.
I pass a complete range of non-null / null (ie a range as long as the values filled with true and false).
The API of the function you show has a list of positions where values are missing.

This is also fine with me to prefer such API but they are 2 different things

Reply:
> This is also fine with me to prefer such API but they are 2 different things

Ah indeed. I think passing a range with the indices of the missing values would be more convenient.

Reply:
> > This is also fine with me to prefer such API but they are 2 different things
> 
> Ah indeed. I think passing a range with the indices of the missing values would be more convenient.

My thinking was the following:
When you have data, do you rather have a list of where the data is missing, or do you have some sort of functor which takes in  a single data point and then decides if its missing or not. For instance once could think of a measurements of floats, and any negative number indicates that something went wrong, then you want to exclude this data point -- in such cases I would prefer to pass an iterator to bools instead of indices)

Reply:
The convenient thing with indices of missing values is that you can provide a default empty argument, and omit it when you know that you don't have any missing values. And building a range of indices from the full bitmap range could be quite easy with the `std::range` functions, I'm not sure the opposite would be as easy (but I might have missed some range constructions).

Reply:
> The convenient thing with indices of missing values is that you can provide a default empty argument, and omit it when you know that you don't have any missing values. 

jeah that's indeed a good argument